staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go

```diff
@@ -207,7 +208,7 @@ func (h *UpgradeAwareHandler) ServeHTTP(w http.ResponseWriter, req *http.Request
 	// Redirect requests with an empty path to a location that ends with a '/'
 	// This is essentially a hack for http://issue.k8s.io/4958.
 	// Note: Keep this code after tryUpgrade to not break that flow.
-	if len(loc.Path) == 0 {
+	if len(loc.Path) == 0 && (method == http.MethodGet || method == http.MethodHead) {
```

Comment:
limiting the redirect to GET/HEAD requests looks good to me.

by continuing with the normal proxy path for other methods (e.g. PUT/POST), what is the HTTP request this change would end up passing to the underlying pod if `len(loc.Path)` was 0 here?

Reply:
From my tests the other methods will pass through to the pod unaffected. Using `kubectl proxy`, `porter` and `curl` I can get 200 OK response back from the pod for the other http methods.
```
$ curl -I -X POST http://localhost:9999/api/v1/namespaces/default/pods/agnhost-7bb97c7fb7-6k9dq/proxy
HTTP/1.1 200 OK
Cache-Control: no-cache, private
Content-Length: 3
Content-Type: text/plain; charset=utf-8
Date: Wed, 30 Sep 2020 18:43:36 GMT
```

Reply:
Using this PR with an updated [e2e test](https://github.com/apisnoop/ticket-writing/blob/NamespacedPodProxy/e2e/proxy.go#L266-L341) for NamespacedPodProxy endpoints gives
```
STEP: Creating a Pod
Oct  1 01:17:33.579: INFO: pod created
Oct  1 01:17:33.591: INFO: Pod Quantity: 1 Status: Pending
Oct  1 01:17:34.596: INFO: Pod Quantity: 1 Status: Pending
Oct  1 01:17:35.597: INFO: Pod Quantity: 1 Status: Pending
Oct  1 01:17:36.596: INFO: Pod Status: Running
Oct  1 01:17:36.596: INFO: Starting http.Client for https://127.0.0.1:41813/api/v1/namespaces/proxy-9664/pods/proxy-pod-target/proxy
Oct  1 01:17:36.600: INFO: http.Client request:GET StatusCode:301
Oct  1 01:17:36.603: INFO: http.Client request:HEAD StatusCode:301
Oct  1 01:17:36.608: INFO: http.Client request:DELETE StatusCode:200
Oct  1 01:17:36.612: INFO: http.Client request:OPTIONS StatusCode:200
Oct  1 01:17:36.616: INFO: http.Client request:PATCH StatusCode:200
Oct  1 01:17:36.621: INFO: http.Client request:POST StatusCode:200
Oct  1 01:17:36.624: INFO: http.Client request:PUT StatusCode:200
```

Reply:
> what is the HTTP request this change would end up passing to the underlying pod if len(loc.Path) was 0 here?

I wanted to make sure the request sent to the underlying pod was a valid HTTP request (especially that a non-empty request-uri was sent to the backing pod)

https://www.w3.org/Protocols/HTTP/1.1/rfc2616bis/draft-lafon-rfc2616bis-03.html#request-uri

> Note that the absolute path cannot be empty; if none is present in the original URI, it must be given as "/" (the server root).

Reply:
- Updated PR so 'absolute path cannot be empty' now.
- Checked how the pod handles the request & payload sent back to the client.
- Tested with `kubectl proxy`, `curl`, `porter` and `tcpdump` 
```
    10.244.0.1.42256 > 10.244.0.5.80: Flags [P.], cksum 0x16e4 (incorrect -> 0x7749), seq 211:419, ack 120, win 502, options [nop,nop,TS val 1754840149 ecr 1101893751], length 208: HTTP, length: 208
        DELETE / HTTP/1.1
        Host: localhost:9999
        User-Agent: via-proxy
        Accept: */*
        Accept-Encoding: gzip
        X-Forwarded-For: 127.0.0.1, 172.18.0.1
        X-Forwarded-Uri: /api/v1/namespaces/default/pods/agnhost2/proxy/

02:58:21.755520 IP (tos 0x0, ttl 64, id 11616, offset 0, flags [DF], proto TCP (6), length 52)
    10.244.0.5.80 > 10.244.0.1.42256: Flags [.], cksum 0x1614 (incorrect -> 0x25f4), ack 419, win 507, options [nop,nop,TS val 1101914757 ecr 1754840149], length 0
02:58:21.755774 IP (tos 0x0, ttl 64, id 11617, offset 0, flags [DF], proto TCP (6), length 171)
    10.244.0.5.80 > 10.244.0.1.42256: Flags [P.], cksum 0x168b (incorrect -> 0x5f0d), seq 120:239, ack 419, win 507, options [nop,nop,TS val 1101914758 ecr 1754840149], length 119: HTTP, length: 119
        HTTP/1.1 200 OK
        Date: Tue, 06 Oct 2020 02:58:21 GMT
        Content-Length: 3
        Content-Type: text/plain; charset=utf-8

        foo[!http]
```

Reply:
Why would we limit to GET & HEAD? Why couldn't people be using POST PUT etc? Isn't this a compatibility breakage?

Reply:
The request body associated with a PUT/POST/PATCH/etc request doesn't get sent to the backend, and the clients I know of drop their PUT/POST/PATCH content in response to a 301 request, and switch to making a GET request to the new location (including [golang](https://github.com/golang/go/blob/0d65c272c9e494cbb604f2bee99d434b8cde46ff/src/net/http/client.go#L505-L516) and [browsers](https://stackoverflow.com/a/13628908/54696), which is what this redirect was originally added for)

That makes returning a 301 redirect for anything other than GET/HEAD lose data and not be usable in the scenarios I can see